### DISCUSSION TITLE
Would this need hairless midi?

### DISCUSSION THREAD

**Discussion by rp1231** [lathoub/Arduino-BLE-MIDI#57] [General]
Hi,
Just wanted to ask if this would also need hairless midi to use or will the bluetooth esp32 be detected as a native midi device?
Thanks

**Comment by lathoub** [comment]
Hi @rp1231 
No need for hairless MIDI, as the ESP32 has native BLE support (or through NimBLE).
See example [here](https://github.com/lathoub/Arduino-BLE-MIDI/blob/master/examples/MidiBle/MidiBle.ino)

**Comment by rp1231** [comment]
Thanks I'll try It out.

**Comment by rp1231** [comment]
Hi,
I'm trying to use 5 rotary encoders with an esp32 over bluetooth.
I'm having trouble figuring out how to use this.
Would you have an example of encoders sending cc messages with esp over bluetooth?

Thanks

**Comment by lathoub** [comment]
IMHO your best bet is to start with the rotary encoders (https://github.com/igorantolic/ai-esp32-rotary-encoder) then use the reading to send MIDI commands.

**Comment by rp1231** [comment]
I had cobbled together some code from an .ino that was meant for the [FortySevenEffects Arduino MIDI Library] but the problem is that the esp connects via ble, but is not detected as a midi device....

Here is a sample test code I tried:
(Not attaching the rotary encoder code for now as it is too big and spans multiple tabs)

```
#include <BLEMIDI_Transport.h>
#include <hardware/BLEMIDI_ESP32.h>
BLEMIDI_CREATE_INSTANCE("E-Drum",MIDI);
  
int tempo = 280;
void setup()
{
  MIDI.begin(10);
}

void loop()
{
    MIDI.sendNoteOn(36, 100, 10);
    MIDI.sendNoteOff(36, 0, 10);
    MIDI.sendNoteOn(44, 100, 10);
    MIDI.sendNoteOff(44, 0, 10);
    delay(tempo);
    
    MIDI.sendNoteOn(44, 100, 10);
    MIDI.sendNoteOff(44, 0, 10);
    delay(tempo);
    
    MIDI.sendNoteOn(38, 100, 10);
    MIDI.sendNoteOff(38, 0, 10);
    MIDI.sendNoteOn(44, 100, 10);
    MIDI.sendNoteOff(44, 0, 10);
    delay(tempo);
    
    MIDI.sendNoteOn(44, 100, 10);
    MIDI.sendNoteOff(44, 0, 10);
    delay(tempo);
}
```
This should work as a test code but isn't working for me.
The esp32 connects to windows as an  unspecified bluetooth device.

**Comment by rp1231** [comment]
Just figured it out.
Turns out windows doesn't support blemidi by default.
I had to install the KORG blemidi driver to get it to work.

**Comment by lathoub** [comment]
Yes, I remember that PITA as well - maybe we should write that down in the wiki section. You want to have a go at that?

**Reply by rp1231** [reply]
Sure, I think it will be helpful to others if it is added to the wiki.
I don't know how to edit the wiki though.....

This is the link to the driver that I used:
https://www.korg.com/us/support/download/driver/1/305/2886/
